<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • enrichOmics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="enrichOmics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">enrichOmics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/Geistlinger_enrichOmics.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/waldronlab/enrichOmics">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Geistlinger_enrichOmics_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      
      <div class="hidden name"><code>Geistlinger_enrichOmics.Rmd</code></div>

    </div>

    
    
<div id="functional-enrichment-analysis-of-high-throughput-omics-data" class="section level1">
<h1 class="hasAnchor">
<a href="#functional-enrichment-analysis-of-high-throughput-omics-data" class="anchor"></a>Functional enrichment analysis of high-throughput omics data</h1>
<div id="workshop-information" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-information" class="anchor"></a>Workshop information</h2>
<div id="instructor-names-and-contact-information" class="section level3">
<h3 class="hasAnchor">
<a href="#instructor-names-and-contact-information" class="anchor"></a>Instructor names and contact information</h3>
<ul>
<li>Ludwig Geistlinger (<a href="mailto:Ludwig.Geistlinger@sph.cuny.edu" class="email">Ludwig.Geistlinger@sph.cuny.edu</a>)</li>
<li>Levi Waldron</li>
</ul>
<p>CUNY School of Public Health 55 W 125th St, New York, NY 10027</p>
</div>
<div id="workshop-description" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h3>
<p>This workshop gives an in-depth overview of existing methods for enrichment analysis of gene expression data with regard to functional gene sets, pathways, and networks. The workshop will help participants understand the distinctions between assumptions and hypotheses of existing methods as well as the differences in objectives and interpretation of results. It will provide code and hands-on practice of all necessary steps for differential expression analysis, gene set- and network-based enrichment analysis, and identification of enriched genomic regions and regulatory elements, along with visualization and exploration of results.</p>
</div>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<ul>
<li><p>Basic knowledge of R syntax</p></li>
<li><p>Familiarity with the SummarizedExperiment class</p></li>
<li><p>Familiarity with the GenomicRanges class</p></li>
<li><p>Familiarity with high-throughput gene expression data as obtained with microarrays and RNA-seq</p></li>
<li><p>Familiarity with the concept of differential expression analysis (with e.g. limma, edgeR, DESeq2)</p></li>
</ul>
</div>
<div id="workshop-participation" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h3>
<p>Execution of example code and hands-on practice</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li><a href="http://bioconductor.org/packages/EnrichmentBrowser">EnrichmentBrowser</a></li>
<li><a href="http://bioconductor.org/packages/regioneR">regioneR</a></li>
<li><a href="http://bioconductor.org/packages/airway">airway</a></li>
<li><a href="http://bioconductor.org/packages/ALL">ALL</a></li>
<li><a href="http://bioconductor.org/packages/hgu95av2.db">hgu95av2.db</a></li>
<li><a href="http://bioconductor.org/packages/BSgenome.Hsapiens.UCSC.hg19.masked">BSgenome.Hsapiens.UCSC.hg19.masked</a></li>
</ul>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Background</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Differential expression analysis</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Gene set analysis</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Gene network analysis</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Genomic region analysis</td>
<td>15m</td>
</tr>
</tbody>
</table>
</div>
<div id="goals-and-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#goals-and-objectives" class="anchor"></a>Goals and objectives</h3>
<p>Theory:</p>
<ul>
<li>Gene sets, pathways &amp; regulatory networks</li>
<li>Resources</li>
<li>Gene set analysis vs. gene set enrichment analysis</li>
<li>Underlying null: competitive vs. self-contained</li>
<li>Generations: ora, fcs &amp; topology-based</li>
</ul>
<p>Practice:</p>
<ul>
<li>Data types: microarray vs. RNA-seq</li>
<li>Differential expression analysis</li>
<li>Defining gene sets according to GO and KEGG</li>
<li>GO/KEGG overrepresentation analysis</li>
<li>Functional class scoring &amp; permutation testing</li>
<li>Network-based enrichment analysis</li>
<li>Genomic region enrichment analysis</li>
</ul>
</div>
<div id="relevant-literature" class="section level3">
<h3 class="hasAnchor">
<a href="#relevant-literature" class="anchor"></a>Relevant literature</h3>
<ul>
<li><p>Geistlinger L, Csaba G, Zimmer R. <a href="https://doi.org/10.1186/s12859-016-0884-1">Bioconductor’s EnrichmentBrowser: seamless navigation through combined results of set- &amp; network-based enrichment analysis</a>. <em>BMC Bioinformatics</em>, 17:45, 2016.</p></li>
<li><p>Geistlinger L, Csaba G, Santarelli M, Ramos M, Schiffer L, Turaga N, Law C, Davis S, Carey V, Morgan M, Zimmer R, Waldron L. <a href="https://doi.org/10.1093/bib/bbz158">Toward a gold standard for benchmarking gene set enrichment analysis</a>. <em>Brief Bioinform</em>, 2020.</p></li>
<li><p>Geistlinger L, Csaba G, Küffner R, Mulder N, Zimmer R. <a href="https://doi.org/10.1093/bioinformatics/btr228">From sets to graphs: towards a realistic enrichment analysis of transcriptomic systems</a>. <em>Bioinformatics</em>, 27(13):i366-73, 2011.</p></li>
<li><p>Geistlinger L, Csaba G, Dirmeier S, Küffner R, Zimmer R. <a href="https://doi.org/10.1093/nar/gkt631">A comprehensive gene regulatory network for the diauxic shift in Saccharomyces cerevisiae</a>. <em>Nucleic Acids Res</em>, 41(18):8452-63, 2013.</p></li>
</ul>
</div>
</div>
<div id="where-does-it-all-come-from" class="section level2">
<h2 class="hasAnchor">
<a href="#where-does-it-all-come-from" class="anchor"></a>Where does it all come from?</h2>
<p>Test whether known biological functions or processes are over-represented (= enriched) in an experimentally-derived gene list, e.g. a list of differentially expressed (DE) genes. See <a href="https://doi.org/10.1093/bioinformatics/btm051">Goeman and Buehlmann, 2007</a> for a critical review.</p>
<p>Example: Transcriptomic study, in which 12,671 genes have been tested for differential expression between two sample conditions and 529 genes were found DE.</p>
<p>Among the DE genes, 28 are annotated to a specific functional gene set, which contains in total 170 genes. This setup corresponds to a 2x2 contingency table,</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="va">deTable</span> <span class="op">&lt;-</span>
     <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">28</span>, <span class="fl">142</span>, <span class="fl">501</span>, <span class="fl">12000</span><span class="op">)</span>,
            nrow <span class="op">=</span> <span class="fl">2</span>,
            dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DE"</span>, <span class="st">"Not.DE"</span><span class="op">)</span>,
                            <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"In.gene.set"</span>, <span class="st">"Not.in.gene.set"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">deTable</span></pre></div>
<pre><code>##        In.gene.set Not.in.gene.set
## DE              28             501
## Not.DE         142           12000</code></pre>
<p>where the overlap of 28 genes can be assessed based on the hypergeometric distribution. This corresponds to a one-sided version of Fisher’s exact test, yielding here a significant enrichment.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/stats/fisher.test.html">fisher.test</a></span><span class="op">(</span><span class="va">deTable</span>, alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></pre></div>
<pre><code>## 
##  Fisher's Exact Test for Count Data
## 
## data:  deTable
## p-value = 4.088e-10
## alternative hypothesis: true odds ratio is greater than 1
## 95 percent confidence interval:
##  3.226736      Inf
## sample estimates:
## odds ratio 
##   4.721744</code></pre>
<p>This basic principle is at the foundation of major public and commercial enrichment tools such as <a href="https://david.ncifcrf.gov/">DAVID</a> and <a href="https://www.pathwaystudio.com">Pathway Studio</a>.</p>
<p>Although gene set enrichment methods have been primarily developed and applied on transcriptomic data, they have recently been modified, extended and applied also in other fields of genomic and biomedical research. This includes novel approaches for functional enrichment analysis of proteomic and metabolomic data as well as genomic regions and disease phenotypes, <a href="https://doi.org/10.1002/0471250953.bi1328s53">Lavallee and Yates, 2016</a>, <a href="https://doi.org/10.1007/978-1-4939-3572-7_20">Chagoyen et al., 2016</a>, <a href="https://doi.org/10.1038/nbt.1630">McLean et al., 2010</a>, <a href="https://doi.org/10.1002/gepi.21617">Ried et al., 2012</a>.</p>
</div>
<div id="gene-expression-based-enrichment-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-expression-based-enrichment-analysis" class="anchor"></a>Gene expression-based enrichment analysis</h2>
<p>The first part of the workshop is largely based on the <a href="http://bioconductor.org/packages/EnrichmentBrowser">EnrichmentBrowser</a> package, which implements an analysis pipeline for high-throughput gene expression data as measured with microarrays and RNA-seq. In a workflow-like manner, the package brings together a selection of established Bioconductor packages for gene expression data analysis. It integrates a wide range of gene set enrichment analysis methods and facilitates combination and exploration of results across methods.</p>
<p><img src="EnrichmentBrowserWorkflow.png" width="100%"></p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">EnrichmentBrowser</span><span class="op">)</span></pre></div>
<p>Further information can be found in the <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/EnrichmentBrowser/inst/doc/EnrichmentBrowser.pdf">vignette</a> and <a href="https://doi.org/10.1186/s12859-016-0884-1">publication</a>.</p>
</div>
<div id="a-primer-on-terminology-existing-methods-statistical-theory" class="section level2">
<h2 class="hasAnchor">
<a href="#a-primer-on-terminology-existing-methods-statistical-theory" class="anchor"></a>A primer on terminology, existing methods &amp; statistical theory</h2>
<p><strong>Gene sets, pathways &amp; regulatory networks</strong></p>
<p>Gene sets are simple lists of usually functionally related genes without further specification of relationships between genes.</p>
<p>Pathways can be interpreted as specific gene sets, typically representing a group of genes that work together in a biological process. Pathways are commonly divided in metabolic and signaling pathways. Metabolic pathways such as glycolysis represent biochemical substrate conversions by specific enzymes. Signaling pathways such as the MAPK signaling pathway describe signal transduction cascades from receptor proteins to transcription factors, resulting in activation or inhibition of specific target genes.</p>
<p>Gene regulatory networks describe the interplay and effects of regulatory factors (such as transcription factors and microRNAs) on the expression of their target genes.</p>
<p><strong>Resources</strong></p>
<p><a href="http://www.geneontology.org">GO</a> and <a href="http://www.genome.jp/kegg">KEGG</a> annotations are most frequently used for the enrichment analysis of functional gene sets. Despite an increasing number of gene set and pathway databases, they are typically the first choice due to their long-standing curation and availability for a wide range of species.</p>
<p><em>GO</em>: The Gene Ontology (GO) consists of three major sub-ontologies that classify gene products according to molecular function (MF), biological process (BP) and cellular component (CC). Each ontology consists of GO terms that define MFs, BPs or CCs to which specific genes are annotated. The terms are organized in a directed acyclic graph, where edges between the terms represent relationships of different types. They relate the terms according to a parent-child scheme, i.e. parent terms denote more general entities, whereas child terms represent more specific entities.</p>
<p><em>KEGG</em>: The Kyoto Encyclopedia of Genes and Genomes (KEGG) is a collection of manually drawn pathway maps representing molecular interaction and reaction networks. These pathways cover a wide range of biochemical processes that can be divided in 7 broad categories: metabolism, genetic and environmental information processing, cellular processes, organismal systems, human diseases, and drug development. Metabolism and drug development pathways differ from pathways of the other 5 categories by illustrating reactions between chemical compounds. Pathways of the other 5 categories illustrate molecular interactions between genes and gene products.</p>
<p><strong>Gene set analysis vs. gene set enrichment analysis</strong></p>
<p>The two predominantly used enrichment methods are:</p>
<ul>
<li>Overrepresentation analysis (ORA), testing whether a gene set contains disproportional many genes of significant expression change, based on the procedure outlined in the first section</li>
<li>Gene set enrichment analysis (GSEA), testing whether genes of a gene set accumulate at the top or bottom of the full gene vector ordered by direction and magnitude of expression change <a href="https://doi.org/10.1073/pnas.0506580102">Subramanian et al., 2005</a>
</li>
</ul>
<p>However, the term <em>gene set enrichment analysis</em> nowadays subsumes a general strategy implemented by a wide range of methods <a href="https://doi.org/10.1093/nar/gkn923">Huang et al., 2009</a>. Those methods have in common the same goal, although approach and statistical model can vary substantially <a href="https://doi.org/10.1093/bioinformatics/btm051">Goeman and Buehlmann, 2007</a>, <a href="https://doi.org/10.1371/journal.pcbi.1002375">Khatri et al., 2012</a>.</p>
<p>To better distinguish from the specific method, some authors use the term <em>gene set analysis</em> to denote the general strategy. However, there is also a specific method from <a href="https://doi.org/10.1214/07-AOAS101">Efron and Tibshirani, 2007</a> of this name.</p>
<p><strong>Underlying null: competitive vs. self-contained</strong></p>
<p><a href="https://doi.org/10.1093/bioinformatics/btm051">Goeman and Buehlmann, 2007</a> classified existing enrichment methods into <em>competitive</em> and <em>self-contained</em> based on the underlying null hypothesis.</p>
<ul>
<li>
<em>Competitive</em> null hypothesis: the genes in the set of interest are at most as often DE as the genes not in the set,</li>
<li>
<em>Self-contained</em> null hypothesis: no genes in the set of interest are DE.</li>
</ul>
<p>Although the authors argue that a self-contained null is closer to the actual question of interest, the vast majority of enrichment methods is competitive.</p>
<p>Goeman and Buehlmann further raise several critical issues concerning the 2x2 ORA:</p>
<ul>
<li>rather arbitrary classification of genes in DE / not DE</li>
<li>based on gene sampling, although sampling of subjects is appropriate</li>
<li>unrealistic independence assumption between genes, resulting in highly anti-conservative <em>p</em>-values</li>
</ul>
<p>With regard to these statistical concerns, GSEA is considered superior:</p>
<ul>
<li>takes all measured genes into account</li>
<li>subject sampling via permutation of class labels<br>
</li>
<li>the incorporated permutation procedure implicitly accounts for correlations between genes</li>
</ul>
<p>However, the simplicity and general applicability of ORA is unmet by subsequent methods improving on these issues. For instance, GSEA requires the expression data as input, which is not available for gene lists derived from other experiment types. On the other hand, the involved sample permutation procedure has been proven inaccurate and time-consuming <a href="https://doi.org/10.1214/07-AOAS101">Efron and Tibshirani, 2007</a>, <a href="https://doi.org/10.2202/1544-6115.1585">Phipson and Smyth, 2010</a>, <a href="https://doi.org/10.1186/s12859-015-0571-7">Larson and Owen, 2015</a>.</p>
<p><strong>Generations: ora, fcs &amp; topology-based</strong></p>
<p><a href="https://doi.org/10.1371/journal.pcbi.1002375">Khatri et al., 2012</a> have taken a slightly different approach by classifying methods along the timeline of development into three generations:</p>
<ol style="list-style-type: decimal">
<li>Generation: ORA methods based on the 2x2 contingency table test,</li>
<li>Generation: functional class scoring (FCS) methods such as GSEA, which compute gene set (= functional class) scores by summarizing per-gene DE statistics,</li>
<li>Generation: topology-based methods, explicitly taking into account interactions between genes as defined in signaling pathways and gene regulatory networks (<a href="https://doi.org/10.1093/bioinformatics/btr228">Geistlinger et al., 2011</a> for an example).</li>
</ol>
<p>Although topology-based (also: network-based) methods appear to be most realistic, their straightforward application can be impaired by features that are not-detectable on the transcriptional level (such as protein-protein interactions) and insufficient network knowledge <a href="https://doi.org/10.1093/nar/gkt631">Geistlinger et al., 2013</a>, <a href="https://doi.org/10.1186/s12859-015-0751-5">Bayerlova et al., 2015</a>.</p>
<p>Given the individual benefits and limitations of existing methods, cautious interpretation of results is required to derive valid conclusions. Whereas no single method is best suited for all application scenarios, applying multiple methods can be beneficial. This has been shown to filter out spurious hits of individual methods, thereby reducing the outcome to gene sets accumulating evidence from different methods <a href="https://doi.org/10.1186/s12859-016-0884-1">Geistlinger et al., 2016</a>, <a href="https://doi.org/10.1093/bioinformatics/btw623">Alhamdoosh et al., 2017</a>.</p>
<p><strong>Guidelines for input and method selection</strong></p>
<p>Based on a comprehensive assessment of 10 major enrichment methods, we recently identified significant differences in runtime and applicability to RNA-seq data, fraction of enriched gene sets depending on the null hypothesis tested, and detection of relevant processes <a href="https://doi.org/10.1093/bib/bbz158">Geistlinger et al., 2020</a>.</p>
<p>Based on these results, we make practical recommendations on how methods originally developed for microarray data can efficiently be applied to RNA-seq data, how to interpret results depending on the type of gene set test conducted and which methods are best suited to effectively prioritize gene sets with high phenotype relevance:</p>
<p>For the exploratory analysis of <em>simple gene lists</em>, we recommend ORA given its ease of applicability, fast runtime and evident relevance of resulting gene set rankings, provided that input gene list and reference gene list are chosen carefully and remembering ORA’s propensity for type I error rate inflation when genes tend to be co-expressed within sets.</p>
<p>For the analysis of <em>pre-ranked gene lists</em> accompanied by gene scores such as fold changes, alternatives to ORA such as <a href="http://bioconductor.org/packages/fgsea">pre-ranked GSEA</a> or <a href="http://web.mit.edu/~r/current/arch/i386_linux26/lib/R/library/limma/html/camera.html">pre-ranked CAMERA</a> exist.</p>
<p>For expression-based enrichment analysis on the <em>full expression matrix</em>, we recommend providing normalized log2 intensities for microarray data, and logTPMs (or logRPKMs / logFPKMs) for RNA-seq data. When given raw read counts, we recommend to apply a variance-stabilizing transformation such as <a href="https://doi.org/10.1186/gb-2014-15-2-r29">voom</a> to arrive at library-size normalized logCPMs.</p>
<p>If the question of interest is to test for association of any gene in the set with the phenotype (<em>self-contained</em> null hypothesis), we recommend ROAST or GSVA that both test a <em>directional</em> hypothesis (genes in the set tend to be either predominantly up- or down-regulated). Both methods can be applied for simple or extended experimental designs, where ROAST is the more natural choice for the comparison of sample groups and also allows one to test a <em>mixed</em> hypothesis (genes in the set tend to be differentially expressed, regardless of the direction). The main strength of GSVA lies in its capabilities for analyzing single samples.</p>
<p>If the question of interest is to test for excess of differential expression in a gene set relative to genes outside the set (<em>competitive</em> null hypothesis), which we believe comes closest to the expectations and intuition of most end users when performing GSEA, we recommend PADOG, which is slower to run but resolves major shortcomings of ORA, and has desirable properties for the analyzed criteria and when compared to other competitive methods. However, PADOG is limited to testing a mixed hypothesis in a comparison of two sample groups, optionally including paired samples or sample batches. Therefore, we recommend the highly customizable SAFE for testing a directional hypothesis or in situations of more complex experimental designs such as comparisons between multiple groups, continuous phenotypes or the presence of covariates.</p>
<p>See also <a href="https://doi.org/10.1093/bib/bbz158">Geistlinger et al., 2020</a> for the results of the benchmarking study and the <a href="http://bioconductor.org/packages/GSEABenchmarkeR">GSEABenchmarkeR</a> package for a general framework for reproducible benchmarking of gene set enrichment methods.</p>
</div>
<div id="data-types" class="section level2">
<h2 class="hasAnchor">
<a href="#data-types" class="anchor"></a>Data types</h2>
<p>Although RNA-seq (read count data) has become the <em>de facto</em> standard for transcriptomic profiling, it is important to know that many methods for differential expression and gene set enrichment analysis have been originally developed for microarray data (intensity measurements).</p>
<p>However, differences in data distribution assumptions (microarray: quasi-normal, RNA-seq: negative binomial) made adaptations in differential expression analysis and, to some extent, also in gene set enrichment analysis necessary.</p>
<p>Thus, we consider two example datasets - a microarray and a RNA-seq dataset, and discuss similarities and differences of the respective analysis steps.</p>
<p>For microarray data, we consider expression measurements of patients with acute lymphoblastic leukemia <a href="https://doi.org/10.1182/blood-2003-09-3243">Chiaretti et al., 2004</a>. A frequent chromosomal defect found among these patients is a translocation, in which parts of chromosome 9 and 22 swap places. This results in the oncogenic fusion gene BCR/ABL created by positioning the ABL1 gene on chromosome 9 to a part of the BCR gene on chromosome 22.</p>
<p>We load the <a href="http://bioconductor.org/packages/ALL">ALL</a> dataset</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ALL</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">ALL</span><span class="op">)</span></pre></div>
<p>and select B-cell ALL patients with and without the BCR/ABL fusion, as described previously <a href="https://www.bioconductor.org/help/publications/books/bioinformatics-and-computational-biology-solutions">Gentleman et al., 2005</a>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">ind.bs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"^B"</span>, <span class="va">ALL</span><span class="op">$</span><span class="va">BT</span><span class="op">)</span>
<span class="va">ind.mut</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">ALL</span><span class="op">$</span><span class="va">mol.biol</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BCR/ABL"</span>, <span class="st">"NEG"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">intersect</a></span><span class="op">(</span><span class="va">ind.bs</span>, <span class="va">ind.mut</span><span class="op">)</span>
<span class="va">all.eset</span> <span class="op">&lt;-</span> <span class="va">ALL</span><span class="op">[</span>, <span class="va">sset</span><span class="op">]</span></pre></div>
<p>We can now access the expression values, which are intensity measurements on a log-scale for 12,625 probes (rows) across 79 patients (columns).</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">all.eset</span><span class="op">)</span></pre></div>
<pre><code>## Features  Samples 
##    12625       79</code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu">exprs</span><span class="op">(</span><span class="va">all.eset</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></pre></div>
<pre><code>##              01005    01010    03002    04007
## 1000_at   7.597323 7.479445 7.567593 7.905312
## 1001_at   5.046194 4.932537 4.799294 4.844565
## 1002_f_at 3.900466 4.208155 3.886169 3.416923
## 1003_s_at 5.903856 6.169024 5.860459 5.687997</code></pre>
<p>As we often have more than one probe per gene, we compute gene expression values as the average of the corresponding probe values.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="va">allSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/probe2gene.html">probe2gene</a></span><span class="op">(</span><span class="va">all.eset</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">allSE</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## [1] "5595" "7075" "1557" "643"  "1843" "4319"</code></pre>
<p>For RNA-seq data, we consider transcriptome profiles of four primary human airway smooth muscle cell lines in two conditions: control and treatment with dexamethasone <a href="https://doi.org/10.1371/journal.pone.0099625">Himes et al., 2014</a>.</p>
<p>We load the <a href="http://bioconductor.org/packages/airway">airway</a> dataset</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></pre></div>
<p>For further analysis, we only keep genes that are annotated to an ENSEMBL gene ID.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="va">airSE</span> <span class="op">&lt;-</span> <span class="va">airway</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"^ENSG"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">airSE</span><span class="op">)</span></pre></div>
<pre><code>## [1] 63677     8</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="fu">assay</span><span class="op">(</span><span class="va">airSE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></pre></div>
<pre><code>##                 SRR1039508 SRR1039509 SRR1039512 SRR1039513
## ENSG00000000003        679        448        873        408
## ENSG00000000005          0          0          0          0
## ENSG00000000419        467        515        621        365
## ENSG00000000457        260        211        263        164</code></pre>
</div>
<div id="differential-expression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-analysis" class="anchor"></a>Differential expression analysis</h2>
<p>Normalization of high-throughput expression data is essential to make results within and between experiments comparable. Microarray (intensity measurements) and RNA-seq (read counts) data typically show distinct features that need to be normalized for. As this is beyond the scope of this workshop, we refer to <a href="http://bioconductor.org/packages/limma">limma</a> for microarray normalization and <a href="http://bioconductor.org/packages/EDASeq">EDASeq</a> for RNA-seq normalization. See also <code><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/normalize.html">EnrichmentBrowser::normalize</a></code>, which wraps commonly used functionality for normalization.</p>
<p>The EnrichmentBrowser incorporates established functionality from the <a href="http://bioconductor.org/packages/limma">limma</a> package for differential expression analysis. This involves the <code>voom</code> transformation when applied to RNA-seq data. Alternatively, differential expression analysis for RNA-seq data can also be carried out based on the negative binomial distribution with <a href="http://bioconductor.org/packages/edgeR">edgeR</a> and <a href="http://bioconductor.org/packages/DESeq2">DESeq2</a>.</p>
<p>This can be performed using the function <code><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/deAna.html">EnrichmentBrowser::deAna</a></code> and assumes some standardized variable names:</p>
<ul>
<li>
<strong>GROUP</strong> defines the sample groups being contrasted,</li>
<li>
<strong>BLOCK</strong> defines paired samples or sample blocks, as e.g. for batch effects.</li>
</ul>
<p>For more information on experimental design, see the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf">limma user’s guide</a>, chapter 9.</p>
<p>For the ALL dataset, the <strong>GROUP</strong> variable indicates whether the BCR-ABL gene fusion is present (1) or not (0).</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="va">allSE</span><span class="op">$</span><span class="va">GROUP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">allSE</span><span class="op">$</span><span class="va">mol.biol</span> <span class="op">==</span> <span class="st">"BCR/ABL"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">allSE</span><span class="op">$</span><span class="va">GROUP</span><span class="op">)</span></pre></div>
<pre><code>## 
##  0  1 
## 42 37</code></pre>
<p>For the airway dataset, it indicates whether the cell lines have been treated with dexamethasone (1) or not (0).</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="va">airSE</span><span class="op">$</span><span class="va">GROUP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">dex</span> <span class="op">==</span> <span class="st">"trt"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">airSE</span><span class="op">$</span><span class="va">GROUP</span><span class="op">)</span></pre></div>
<pre><code>## 
## 0 1 
## 4 4</code></pre>
<p>Paired samples, or in general sample batches/blocks, can be defined via a <code>BLOCK</code> column in the <code>colData</code> slot. For the airway dataset, the sample blocks correspond to the four different cell lines.</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="va">airSE</span><span class="op">$</span><span class="va">BLOCK</span> <span class="op">&lt;-</span> <span class="va">airway</span><span class="op">$</span><span class="va">cell</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">airSE</span><span class="op">$</span><span class="va">BLOCK</span><span class="op">)</span></pre></div>
<pre><code>## 
## N052611 N061011 N080611  N61311 
##       2       2       2       2</code></pre>
<p>For microarray data, the <code><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/deAna.html">EnrichmentBrowser::deAna</a></code> function carries out differential expression analysis based on functionality from the <em>limma</em> package. Resulting log2 fold changes and <em>t</em>-test derived <em>p</em>-values for each gene are appended to the <code>rowData</code> slot.</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="va">allSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/deAna.html">deAna</a></span><span class="op">(</span><span class="va">allSE</span><span class="op">)</span>
<span class="fu">rowData</span><span class="op">(</span><span class="va">allSE</span><span class="op">)</span></pre></div>
<pre><code>## DataFrame with 9009 rows and 4 columns
##               FC limma.STAT      PVAL  ADJ.PVAL
##        &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## 5595   0.0429699   0.734676 0.4646865  0.899340
## 7075   0.0320835   0.454690 0.6505647  0.949052
## 1557  -0.0439401  -1.065773 0.2897348  0.818508
## 643   -0.0277544  -0.567391 0.5720386  0.929215
## 1843  -0.4273025  -1.750504 0.0838684  0.566394
## ...          ...        ...       ...       ...
## 6300 -0.02665177  -0.608604  0.544512  0.922900
## 7297 -0.12425768  -1.112795  0.269132  0.804769
## 2246  0.05224289   1.274073  0.206329  0.748238
## 7850 -0.00908230  -0.102406  0.918691  0.991836
## 1593 -0.00747714  -0.145650  0.884565  0.989546</code></pre>
<p>Nominal <em>p</em>-values are already corrected for multiple testing (<code>ADJ.PVAL</code>) using the method from Benjamini and Hochberg implemented in <code><a href="https://rdrr.io/r/stats/p.adjust.html">stats::p.adjust</a></code>.</p>
<p>For RNA-seq data, the <code>deAna</code> function can be used to carry out differential expression analysis between the two groups either based on functionality from <em>limma</em> (that includes the <code>voom</code> transformation), or alternatively, the frequently used <em>edgeR</em> or <em>DESeq2</em> package. Here, we use the analysis based on <em>edgeR</em>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="va">airSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/deAna.html">deAna</a></span><span class="op">(</span><span class="va">airSE</span>, de.method <span class="op">=</span> <span class="st">"edgeR"</span><span class="op">)</span></pre></div>
<pre><code>## Excluding 47751 genes not satisfying min.cpm threshold</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="fu">rowData</span><span class="op">(</span><span class="va">airSE</span><span class="op">)</span></pre></div>
<pre><code>## DataFrame with 15926 rows and 4 columns
##                         FC edgeR.STAT        PVAL   ADJ.PVAL
##                  &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt;
## ENSG00000000003 -0.3901002 31.0558140 0.000232422 0.00217355
## ENSG00000000419  0.1978022  6.6454709 0.027419893 0.07560513
## ENSG00000000457  0.0291609  0.0929623 0.766666551 0.84808859
## ENSG00000000460 -0.1243820  0.3832263 0.549659194 0.67996523
## ENSG00000000971  0.4172901 28.7686093 0.000312276 0.00272063
## ...                    ...        ...         ...        ...
## ENSG00000273373 -0.0438722  0.0397087   0.8460260   0.901607
## ENSG00000273382 -0.8597567  7.7869742   0.0190219   0.057267
## ENSG00000273448  0.0281667  0.0103270   0.9752405   0.984888
## ENSG00000273472 -0.4642705  1.9010366   0.1978818   0.328963
## ENSG00000273486 -0.1109445  0.1536285   0.7032766   0.802377</code></pre>
<p><em>Exercise:</em> Compare the number of differentially expressed genes as obtained on the <code>airSE</code> with <code>limma/voom</code>, <code>edgeR</code>, and <code>DESeq2</code>.</p>
</div>
<div id="gene-sets" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-sets" class="anchor"></a>Gene sets</h2>
<p>We are now interested in whether pre-defined sets of genes that are known to work together, e.g. as defined in the <a href="http://www.geneontology.org">Gene Ontology</a> or the <a href="http://www.genome.jp/kegg">KEGG</a> pathway annotation, are coordinately differentially expressed.</p>
<p>The function <code>getGenesets</code> can be used to download gene sets from databases such as GO and KEGG. Here, we use the function to download all KEGG pathways for a chosen organism (here: ) as gene sets.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="va">kegg.gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/getGenesets.html">getGenesets</a></span><span class="op">(</span>org <span class="op">=</span> <span class="st">"hsa"</span>, db <span class="op">=</span> <span class="st">"kegg"</span><span class="op">)</span>
<span class="va">kegg.gs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></pre></div>
<pre><code>## $`hsa00010_Glycolysis_/_Gluconeogenesis`
##  [1] "10327"  "124"    "125"    "126"    "127"    "128"    "130"    "130589"
##  [9] "131"    "160287" "1737"   "1738"   "2023"   "2026"   "2027"   "217"   
## [17] "218"    "219"    "2203"   "221"    "222"    "223"    "224"    "226"   
## [25] "229"    "230"    "2538"   "2597"   "26330"  "2645"   "2821"   "3098"  
## [33] "3099"   "3101"   "387712" "3939"   "3945"   "3948"   "441531" "501"   
## [41] "5105"   "5106"   "5160"   "5161"   "5162"   "5211"   "5213"   "5214"  
## [49] "5223"   "5224"   "5230"   "5232"   "5236"   "5313"   "5315"   "55276" 
## [57] "55902"  "57818"  "669"    "7167"   "80201"  "83440"  "84532"  "8789"  
## [65] "92483"  "92579"  "9562"  
## 
## $`hsa00020_Citrate_cycle_(TCA_cycle)`
##  [1] "1431"  "1737"  "1738"  "1743"  "2271"  "3417"  "3418"  "3419"  "3420" 
## [10] "3421"  "4190"  "4191"  "47"    "48"    "4967"  "50"    "5091"  "5105" 
## [19] "5106"  "5160"  "5161"  "5162"  "55753" "6389"  "6390"  "6391"  "6392" 
## [28] "8801"  "8802"  "8803"</code></pre>
<p>Analogously, the function <code>getGenesets</code> can be used to retrieve GO terms of a selected ontology (here: biological process, BP) as defined in the <em>GO.db</em> annotation package.</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="va">go.gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/getGenesets.html">getGenesets</a></span><span class="op">(</span>org <span class="op">=</span> <span class="st">"hsa"</span>, db <span class="op">=</span> <span class="st">"go"</span>, onto <span class="op">=</span> <span class="st">"BP"</span>, mode <span class="op">=</span> <span class="st">"GO.db"</span><span class="op">)</span>
<span class="va">go.gs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></pre></div>
<pre><code>## $`GO:0000002_mitochondrial_genome_maintenance`
##  [1] "10000" "1890"  "291"   "4205"  "4358"  "4976"  "55154" "55186" "80119"
## [10] "84275" "92667" "9361" 
## 
## $`GO:0000003_reproduction`
## [1] "2796"   "2797"   "286826" "8510"</code></pre>
<p>If provided a file, the function <code>getGenesets</code> parses user-defined gene sets from GMT file format. Here, we use this functionality for reading a list of already downloaded KEGG gene sets for <em>Homo sapiens</em> containing NCBI Entrez Gene IDs.</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="va">data.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"EnrichmentBrowser"</span><span class="op">)</span>
<span class="va">gmt.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">data.dir</span>, <span class="st">"hsa_kegg_gs.gmt"</span><span class="op">)</span>
<span class="va">hsa.gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/getGenesets.html">getGenesets</a></span><span class="op">(</span><span class="va">gmt.file</span><span class="op">)</span>
<span class="va">hsa.gs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></pre></div>
<pre><code>## $hsa05416_Viral_myocarditis
##  [1] "100509457" "101060835" "1525"      "1604"      "1605"      "1756"     
##  [7] "1981"      "1982"      "25"        "2534"      "27"        "3105"     
## [13] "3106"      "3107"      "3108"      "3109"      "3111"      "3112"     
## [19] "3113"      "3115"      "3117"      "3118"      "3119"      "3122"     
## [25] "3123"      "3125"      "3126"      "3127"      "3133"      "3134"     
## [31] "3135"      "3383"      "3683"      "3689"      "3908"      "4624"     
## [37] "4625"      "54205"     "5551"      "5879"      "5880"      "5881"     
## [43] "595"       "60"        "637"       "6442"      "6443"      "6444"     
## [49] "6445"      "71"        "836"       "841"       "842"       "857"      
## [55] "8672"      "940"       "941"       "942"       "958"       "959"      
## 
## $`hsa04622_RIG-I-like_receptor_signaling_pathway`
##  [1] "10010"  "1147"   "1432"   "1540"   "1654"   "23586"  "26007"  "29110" 
##  [9] "338376" "340061" "3439"   "3440"   "3441"   "3442"   "3443"   "3444"  
## [17] "3445"   "3446"   "3447"   "3448"   "3449"   "3451"   "3452"   "3456"  
## [25] "3467"   "3551"   "3576"   "3592"   "3593"   "3627"   "3661"   "3665"  
## [33] "4214"   "4790"   "4792"   "4793"   "5300"   "54941"  "55593"  "5599"  
## [41] "5600"   "5601"   "5602"   "5603"   "56832"  "57506"  "5970"   "6300"  
## [49] "64135"  "64343"  "6885"   "7124"   "7186"   "7187"   "7189"   "7706"  
## [57] "79132"  "79671"  "80143"  "841"    "843"    "8517"   "8717"   "8737"  
## [65] "8772"   "9140"   "9474"   "9636"   "9641"   "9755"</code></pre>
<p>Note #1: Gene set collections for 11 different species from the<br><a href="http://software.broadinstitute.org/gsea/msigdb/collections.jsp">Molecular Signatures Database (MSigDB)</a> can be obtained using <code>getGenesets</code> with <code>db = "msigdb"</code>. For example, the <em>H</em>allmark gene set collection can be obtained from MSigDB via:</p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="va">hall.gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/getGenesets.html">getGenesets</a></span><span class="op">(</span>org <span class="op">=</span> <span class="st">"hsa"</span>, db <span class="op">=</span> <span class="st">"msigdb"</span>, cat <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span> 
<span class="va">hall.gs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></pre></div>
<pre><code>## $M5890_HALLMARK_TNFA_SIGNALING_VIA_NFKB
##   [1] "19"     "57007"  "374"    "467"    "490"    "2683"   "9334"   "597"   
##   [9] "602"    "604"    "8553"   "329"    "330"    "650"    "694"    "7832"  
##  [17] "10950"  "6347"   "6364"   "6351"   "6352"   "3491"   "595"    "57018" 
##  [25] "9034"   "960"    "969"    "941"    "9308"   "1026"   "1051"   "1052"  
##  [33] "8837"   "23529"  "1435"   "1437"   "2919"   "3627"   "6373"   "2920"  
##  [41] "2921"   "6372"   "23586"  "23258"  "11080"  "55332"  "1843"   "1844"  
##  [49] "1846"   "1847"   "1906"   "1942"   "1958"   "1959"   "1960"   "10938" 
##  [57] "10209"  "2114"   "2150"   "2152"   "24147"  "2353"   "2354"   "8061"  
##  [65] "2355"   "2526"   "50486"  "1647"   "4616"   "2643"   "2669"   "9945"  
##  [73] "1880"   "1839"   "3280"   "3383"   "23308"  "3398"   "9592"   "8870"  
##  [81] "51278"  "64135"  "3433"   "3460"   "3593"   "3601"   "3606"   "3552"  
##  [89] "3553"   "51561"  "3569"   "3572"   "3575"   "3624"   "3659"   "8660"  
##  [97] "182"    "3725"   "3726"   "23135"  "7071"   "10365"  "9314"   "1316"  
## [105] "687"    "8942"   "3914"   "3949"   "3976"   "9516"   "23764"  "5606"  
## [113] "1326"   "4082"   "4170"   "9242"   "4084"   "4609"   "10135"  "10725" 
## [121] "4780"   "4783"   "4790"   "4791"   "4792"   "4794"   "4814"   "3164"  
## [129] "4929"   "8013"   "4973"   "24145"  "5142"   "10611"  "5187"   "5209"  
## [137] "22822"  "7262"   "5328"   "5329"   "5341"   "10769"  "8613"   "56937" 
## [145] "10957"  "23645"  "5734"   "5743"   "5791"   "5806"   "1827"   "5966"  
## [153] "5970"   "5971"   "388"    "8767"   "127544" "6303"   "6385"   "5055"  
## [161] "5271"   "5054"   "6446"   "150094" "9120"   "6515"   "11182"  "4088"  
## [169] "8303"   "9021"   "6648"   "8877"   "80176"  "8878"   "6776"   "10010" 
## [177] "6890"   "7050"   "25976"  "7097"   "3371"   "7124"   "7127"   "7128"  
## [185] "7130"   "25816"  "3604"   "8744"   "10318"  "79155"  "7185"   "10221" 
## [193] "9322"   "8848"   "7280"   "7422"   "79693"  "65986"  "80149"  "7538"  
## 
## $M5891_HALLMARK_HYPOXIA
##   [1] "57007"  "133"    "136"    "205"    "9590"   "226"    "229"    "230"   
##   [9] "272"    "51129"  "55139"  "302"    "467"    "538"    "126792" "124872"
##  [17] "63827"  "596"    "633"    "8553"   "665"    "680"    "694"    "771"   
##  [25] "839"    "857"    "284119" "112464" "3491"   "1490"   "8839"   "901"   
##  [33] "1026"   "1027"   "1028"   "9435"   "9469"   "10370"  "1289"   "1356"  
##  [41] "1466"   "7852"   "1634"   "1649"   "54541"  "10570"  "1837"   "1843"  
##  [49] "1907"   "1942"   "1944"   "1956"   "2023"   "2026"   "2027"   "30001" 
##  [57] "54206"  "2113"   "2131"   "2152"   "26355"  "2203"   "2353"   "2355"  
##  [65] "2309"   "2548"   "2584"   "2597"   "26330"  "2632"   "2645"   "2651"  
##  [73] "2745"   "2817"   "2719"   "2239"   "2821"   "9380"   "2997"   "3036"  
##  [81] "3069"   "3073"   "3098"   "3099"   "3162"   "3219"   "9957"   "3309"  
##  [89] "3423"   "8870"   "3484"   "3486"   "3569"   "10994"  "3623"   "8660"  
##  [97] "3669"   "23210"  "3725"   "11015"  "55818"  "3798"   "1316"   "8609"  
## [105] "54800"  "3906"   "9215"   "3939"   "3948"   "4015"   "56925"  "23764" 
## [113] "4214"   "4282"   "4493"   "4502"   "4601"   "4627"   "55577"  "1463"  
## [121] "10397"  "3340"   "8509"   "23327"  "4783"   "25819"  "2908"   "5033"  
## [129] "8974"   "5066"   "5105"   "5155"   "5163"   "5165"   "5209"   "5211"  
## [137] "5214"   "5224"   "5228"   "5230"   "5236"   "55276"  "5260"   "5292"  
## [145] "5313"   "5317"   "51316"  "5329"   "123"    "10957"  "10891"  "8497"  
## [153] "23645"  "5507"   "25824"  "5578"   "5837"   "3516"   "6095"   "58528" 
## [161] "6275"   "8819"   "949"    "6383"   "9672"   "6385"   "8991"   "5054"  
## [169] "6478"   "6576"   "6513"   "6515"   "6518"   "2542"   "6533"   "8406"  
## [177] "8987"   "6781"   "8614"   "6820"   "26136"  "7043"   "7045"   "7052"  
## [185] "25976"  "8277"   "55076"  "7128"   "7162"   "7163"   "7167"   "8459"  
## [193] "7360"   "7422"   "7428"   "7436"   "26118"  "7511"   "7538"   "23036"</code></pre>
<p>Note #2: Gene set libraries from the comprehensive <a href="https://amp.pharm.mssm.edu/Enrichr/#stats">Enrichr</a> collection for 5 different species can be obtained using <code>getGenesets</code> with <code>db = "enrichr"</code>. For example, gene sets based on transcription factor interactions can be obtained from Enrichr via:</p>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="va">tfppi.gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/getGenesets.html">getGenesets</a></span><span class="op">(</span>org <span class="op">=</span> <span class="st">"hsa"</span>, db <span class="op">=</span> <span class="st">"enrichr"</span>, lib <span class="op">=</span> <span class="st">"Transcription_Factor_PPIs"</span><span class="op">)</span>
<span class="va">tfppi.gs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></pre></div>
<p>Note #3: The <code>idMap</code> function can be used to map gene sets from NCBI Entrez Gene IDs to other common gene ID types such as ENSEMBL gene IDs or HGNC symbols.\ For example, to map the gene sets from Entrez Gene IDs to gene symbols:</p>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="va">hsa.gs.sym</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/idMap.html">idMap</a></span><span class="op">(</span><span class="va">hsa.gs</span>, org <span class="op">=</span> <span class="st">"hsa"</span>, from <span class="op">=</span> <span class="st">"ENTREZID"</span>, to <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span>
<span class="va">hsa.gs.sym</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></pre></div>
<pre><code>## $hsa05416_Viral_myocarditis
##  [1] "ABL1"     "ABL2"     "ACTB"     "ACTG1"    "BID"      "CASP3"   
##  [7] "CASP8"    "CASP9"    "CAV1"     "CCND1"    "CD28"     "CD40"    
## [13] "CD40LG"   "CD55"     "CD80"     "CD86"     "CXADR"    "CYCS"    
## [19] "DAG1"     "DMD"      "EIF4G1"   "EIF4G2"   "EIF4G3"   "FYN"     
## [25] "HLA-A"    "HLA-B"    "HLA-C"    "HLA-DMA"  "HLA-DMB"  "HLA-DOA" 
## [31] "HLA-DOB"  "HLA-DPA1" "HLA-DPB1" "HLA-DQA1" "HLA-DQA2" "HLA-DQB1"
## [37] "HLA-DRA"  "HLA-DRB1" "HLA-DRB3" "HLA-DRB4" "HLA-DRB5" "HLA-E"   
## [43] "HLA-F"    "HLA-G"    "ICAM1"    "ITGAL"    "ITGB2"    "LAMA2"   
## [49] "MYH6"     "MYH7"     "PRF1"     "RAC1"     "RAC2"     "RAC3"    
## [55] "SGCA"     "SGCB"     "SGCD"     "SGCG"    
## 
## $`hsa04622_RIG-I-like_receptor_signaling_pathway`
##  [1] "ATG12"  "ATG5"   "AZI2"   "CASP10" "CASP8"  "CHUK"   "CXCL10" "CXCL8" 
##  [9] "CYLD"   "DDX3X"  "DDX58"  "DHX58"  "FADD"   "IFIH1"  "IFNA1"  "IFNA10"
## [17] "IFNA13" "IFNA14" "IFNA16" "IFNA17" "IFNA2"  "IFNA21" "IFNA4"  "IFNA5" 
## [25] "IFNA6"  "IFNA7"  "IFNA8"  "IFNB1"  "IFNE"   "IFNK"   "IFNW1"  "IKBKB" 
## [33] "IKBKE"  "IKBKG"  "IL12A"  "IL12B"  "IRF3"   "IRF7"   "ISG15"  "MAP3K1"
## [41] "MAP3K7" "MAPK10" "MAPK11" "MAPK12" "MAPK13" "MAPK14" "MAPK8"  "MAPK9" 
## [49] "MAVS"   "NFKB1"  "NFKBIA" "NFKBIB" "NLRX1"  "OTUD5"  "PIN1"   "RELA"  
## [57] "RIPK1"  "RNF125" "SIKE1"  "STING1" "TANK"   "TBK1"   "TBKBP1" "TKFC"  
## [65] "TNF"    "TRADD"  "TRAF2"  "TRAF3"  "TRAF6"  "TRIM25"</code></pre>
</div>
<div id="gokegg-overrepresentation-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#gokegg-overrepresentation-analysis" class="anchor"></a>GO/KEGG overrepresentation analysis</h2>
<p>A variety of gene set analysis methods have been proposed <a href="https://doi.org/10.1371/journal.pcbi.1002375">Khatri et al., 2012</a>. The most basic, yet frequently used, method is the over-representation analysis (ORA) with gene sets defined according to GO or KEGG. As outlined in the first section, ORA tests the overlap between DE genes (typically DE <em>p</em>-value &lt; 0.05) and genes in a gene set based on the hypergeometric distribution. Here, we choose a significance level <span class="math inline">\(\alpha = 0.2\)</span> for demonstration.</p>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="va">ora.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/sbea.html">sbea</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"ora"</span>, se <span class="op">=</span> <span class="va">allSE</span>, gs <span class="op">=</span> <span class="va">hsa.gs</span>, perm <span class="op">=</span> <span class="fl">0</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/eaBrowse.html">gsRanking</a></span><span class="op">(</span><span class="va">ora.all</span><span class="op">)</span></pre></div>
<pre><code>## DataFrame with 7 rows and 4 columns
##                 GENE.SET  NR.GENES NR.SIG.GENES      PVAL
##              &lt;character&gt; &lt;numeric&gt;    &lt;numeric&gt; &lt;numeric&gt;
## 1 hsa05202_Transcripti..       153           17    0.0351
## 2 hsa05412_Arrhythmoge..        63            8    0.0717
## 3       hsa05144_Malaria        45            6    0.0932
## 4 hsa04670_Leukocyte_t..        94           10    0.1220
## 5 hsa05100_Bacterial_i..        64            7    0.1620
## 6 hsa04622_RIG-I-like_..        54            6    0.1780
## 7 hsa05130_Pathogenic_..        43            5    0.1840</code></pre>
<p>Such a ranked list is the standard output of most existing enrichment tools. Using the <code>eaBrowse</code> function creates a HTML summary from which each gene set can be inspected in more detail.</p>
<p><img src="EnrichmentBrowserNavigation.png" width="100%"></p>
<div class="sourceCode" id="cb44"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/eaBrowse.html">eaBrowse</a></span><span class="op">(</span><span class="va">ora.all</span><span class="op">)</span></pre></div>
<p>The resulting summary page includes for each significant gene set</p>
<ul>
<li>a gene report, which lists all genes of a set along with fold change and DE <span class="math inline">\(p\)</span>-value (click on links in column <code>NR.GENES</code>),</li>
<li>interactive overview plots such as heatmap and volcano plot (column <code>SET.VIEW</code>, supports mouse-over and click-on),</li>
<li>for KEGG pathways: highlighting of differentially expressed genes on the pathway maps (column <code>PATH.VIEW</code>, supports mouse-over and click-on).</li>
</ul>
<p>As ORA works on the list of DE genes and not the actual expression values, it can be straightforward applied to RNA-seq data. However, as the gene sets here contain NCBI Entrez gene IDs and the airway dataset contains ENSEMBL gene ids, we first map the airway dataset to Entrez IDs.</p>
<div class="sourceCode" id="cb45"><pre class="downlit">
<span class="va">airSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/idMap.html">idMap</a></span><span class="op">(</span><span class="va">airSE</span>, org <span class="op">=</span> <span class="st">"hsa"</span>, from <span class="op">=</span> <span class="st">"ENSEMBL"</span>, to <span class="op">=</span> <span class="st">"ENTREZID"</span><span class="op">)</span></pre></div>
<pre><code>## Encountered 85 from.IDs with &gt;1 corresponding to.ID</code></pre>
<pre><code>## (the first to.ID was chosen for each of them)</code></pre>
<pre><code>## Excluded 2391 from.IDs without a corresponding to.ID</code></pre>
<pre><code>## Encountered 6 to.IDs with &gt;1 from.ID (the first from.ID was chosen for each of them)</code></pre>
<pre><code>## Mapped from.IDs have been added to the rowData column ENSEMBL</code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit">
<span class="va">ora.air</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/sbea.html">sbea</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"ora"</span>, se <span class="op">=</span> <span class="va">airSE</span>, gs <span class="op">=</span> <span class="va">go.gs</span>, perm <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/eaBrowse.html">gsRanking</a></span><span class="op">(</span><span class="va">ora.air</span><span class="op">)</span></pre></div>
<pre><code>## DataFrame with 360 rows and 4 columns
##                   GENE.SET  NR.GENES NR.SIG.GENES      PVAL
##                &lt;character&gt; &lt;numeric&gt;    &lt;numeric&gt; &lt;numeric&gt;
## 1   GO:0030198_extracell..       163           96  7.87e-10
## 2   GO:0007155_cell_adhe..       291          149  2.08e-08
## 3   GO:0030336_negative_..        88           56  6.71e-08
## 4   GO:0007186_G_protein..       282          136  6.11e-06
## 5   GO:0006936_muscle_co..        62           38  3.03e-05
## ...                    ...       ...          ...       ...
## 356 GO:0033077_T_cell_di..        24           13    0.0472
## 357 GO:0048839_inner_ear..        24           13    0.0472
## 358 GO:0030282_bone_mine..        31           16    0.0480
## 359 GO:0031100_animal_or..        31           16    0.0480
## 360 GO:0051781_positive_..        31           16    0.0480</code></pre>
<p>Note #1: <a href="https://doi.org/10.1186/gb-2010-11-2-r14">Young et al., 2010</a>, have reported biased results for ORA on RNA-seq data due to over-detection of differential expression for long and highly expressed transcripts. The <a href="http://bioconductor.org/packages/goseq">goseq</a> package and <code><a href="https://rdrr.io/pkg/limma/man/goana.html">limma::goana</a></code> implement possibilities to adjust ORA for gene length and abundance bias.</p>
<p>Note #2: Independent of the expression data type under investigation, overlap between gene sets can result in redundant findings. This is well-documented for GO (parent-child structure, <a href="https://doi.org/10.1038/nrg2363">Rhee et al., 2008</a>) and KEGG (pathway overlap/crosstalk, <a href="https://doi.org/10.1101/gr.153551.112">Donato et al., 2013</a>). The <a href="http://bioconductor.org/packages/topGO">topGO</a> package (explicitly designed for GO) and <a href="http://bioconductor.org/packages/mgsa">mgsa</a> (applicable to arbitrary gene set definitions) implement modifications of ORA to account for such redundancies.</p>
</div>
<div id="functional-class-scoring-permutation-testing" class="section level2">
<h2 class="hasAnchor">
<a href="#functional-class-scoring-permutation-testing" class="anchor"></a>Functional class scoring &amp; permutation testing</h2>
<p>A major limitation of ORA is that it restricts analysis to DE genes, excluding genes not satisfying the chosen significance threshold (typically the vast majority).</p>
<p>This is resolved by gene set enrichment analysis (GSEA), which scores the tendency of gene set members to appear rather at the top or bottom of the ranked list of all measured genes <a href="https://doi.org/10.1073/pnas.0506580102">Subramanian et al., 2005</a>. The statistical significance of the enrichment score (ES) of a gene set is assessed via sample permutation, i.e. (1) sample labels (= group assignment) are shuffled, (2) per-gene DE statistics are recomputed, and (3) the enrichment score is recomputed. Repeating this procedure many times allows to determine the empirical distribution of the enrichment score and to compare the observed enrichment score against it. Here, we carry out GSEA with 1000 permutations.</p>
<div class="sourceCode" id="cb53"><pre class="downlit">
<span class="va">gsea.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/sbea.html">sbea</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"gsea"</span>, se <span class="op">=</span> <span class="va">allSE</span>, gs <span class="op">=</span> <span class="va">hsa.gs</span>, perm <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>  </pre></div>
<pre><code>## Permutations: 1 -- 100</code></pre>
<pre><code>## Permutations: 101 -- 200</code></pre>
<pre><code>## Permutations: 201 -- 300</code></pre>
<pre><code>## Permutations: 301 -- 400</code></pre>
<pre><code>## Permutations: 401 -- 500</code></pre>
<pre><code>## Permutations: 501 -- 600</code></pre>
<pre><code>## Permutations: 601 -- 700</code></pre>
<pre><code>## Permutations: 701 -- 800</code></pre>
<pre><code>## Permutations: 801 -- 900</code></pre>
<pre><code>## Permutations: 901 -- 1000</code></pre>
<pre><code>## Processing ...</code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/eaBrowse.html">gsRanking</a></span><span class="op">(</span><span class="va">gsea.all</span><span class="op">)</span></pre></div>
<pre><code>## DataFrame with 19 rows and 4 columns
##                   GENE.SET        ES       NES      PVAL
##                &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## 1   hsa05412_Arrhythmoge..     0.511      1.86   0.00000
## 2   hsa04520_Adherens_ju..     0.488      1.73   0.00000
## 3   hsa05202_Transcripti..     0.530      1.72   0.00000
## 4   hsa05416_Viral_myoca..     0.630      1.89   0.00191
## 5   hsa04670_Leukocyte_t..     0.499      1.79   0.00194
## ...                    ...       ...       ...       ...
## 15  hsa04621_NOD-like_re..     0.633      1.56    0.0216
## 16  hsa05150_Staphylococ..     0.576      1.69    0.0231
## 17    hsa05131_Shigellosis     0.479      1.52    0.0236
## 18  hsa05130_Pathogenic_..     0.486      1.52    0.0283
## 19      hsa04210_Apoptosis     0.424      1.42    0.0413</code></pre>
<p>As GSEA’s permutation procedure involves re-computation of per-gene DE statistics, adaptations are necessary for RNA-seq. When analyzing RNA-seq datasets with expression values given as logTPMs (or logRPKMs / logFPKMs), the available set-based enrichment methods can be applied as for microarray data. However, when given raw read counts as for the airway dataset, we recommend to first apply a variance-stabilizing transformation such as <code>voom</code> to arrive at library-size normalized logCPMs.</p>
<div class="sourceCode" id="cb67"><pre class="downlit">
<span class="va">airSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/normalize.html">normalize</a></span><span class="op">(</span><span class="va">airSE</span>, norm.method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></pre></div>
<p>The mean-variance relationship of the transformed data is similar to what is observed for microarray data, simplifying the application of legacy enrichment methods such as GSEA and PADOG to RNA-seq data, and enable the use of fast and established methods.</p>
<div class="sourceCode" id="cb68"><pre class="downlit">
<span class="va">gsea.air</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/sbea.html">sbea</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"gsea"</span>, se <span class="op">=</span> <span class="va">airSE</span>, gs <span class="op">=</span> <span class="va">kegg.gs</span><span class="op">)</span>  </pre></div>
<pre><code>## Permutations: 1 -- 100</code></pre>
<pre><code>## Permutations: 101 -- 200</code></pre>
<pre><code>## Permutations: 201 -- 300</code></pre>
<pre><code>## Permutations: 301 -- 400</code></pre>
<pre><code>## Permutations: 401 -- 500</code></pre>
<pre><code>## Permutations: 501 -- 600</code></pre>
<pre><code>## Permutations: 601 -- 700</code></pre>
<pre><code>## Permutations: 701 -- 800</code></pre>
<pre><code>## Permutations: 801 -- 900</code></pre>
<pre><code>## Permutations: 901 -- 1000</code></pre>
<pre><code>## Processing ...</code></pre>
<p>While it might be in some cases necessary to apply permutation-based GSEA for RNA-seq data, there are also alternatives avoiding permutation. Among them is ROtAtion gene Set Testing (ROAST), which uses rotation instead of permutation <a href="https://doi.org/10.1093/bioinformatics/btq401">Wu et al., 2010</a>.</p>
<div class="sourceCode" id="cb80"><pre class="downlit">
<span class="va">roast.air</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/sbea.html">sbea</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"roast"</span>, se <span class="op">=</span> <span class="va">airSE</span>, gs <span class="op">=</span> <span class="va">hsa.gs</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/eaBrowse.html">gsRanking</a></span><span class="op">(</span><span class="va">roast.air</span><span class="op">)</span>  </pre></div>
<pre><code>## DataFrame with 23 rows and 4 columns
##                   GENE.SET  NR.GENES       DIR      PVAL
##                &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## 1     hsa05131_Shigellosis        56         1  0.000999
## 2   hsa05206_MicroRNAs_i..       129        -1  0.000999
## 3   hsa03030_DNA_replica..        35        -1  0.000999
## 4   hsa05323_Rheumatoid_..        52        -1  0.002000
## 5   hsa04390_Hippo_signa..       116         1  0.003000
## ...                    ...       ...       ...       ...
## 19  hsa00340_Histidine_m..        15         1     0.019
## 20  hsa04150_mTOR_signal..        51         1     0.020
## 21       hsa05218_Melanoma        52        -1     0.020
## 22  hsa04514_Cell_adhesi..        74        -1     0.025
## 23  hsa04550_Signaling_p..       101         1     0.038</code></pre>
<p>A selection of additional methods is also available:</p>
<div class="sourceCode" id="cb82"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/sbea.html">sbeaMethods</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<pre><code>##  [1] "ora"        "safe"       "gsea"       "gsa"        "padog"     
##  [6] "globaltest" "roast"      "camera"     "gsva"       "samgs"     
## [11] "ebm"        "mgsa"</code></pre>
<p><em>Exercise</em>: Carry out a GO overrepresentation analysis for the <code>allSE</code> and <code>airSE</code>. How many significant gene sets do you observe in each case?</p>
</div>
<div id="network-based-enrichment-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#network-based-enrichment-analysis" class="anchor"></a>Network-based enrichment analysis</h2>
<p>Having found gene sets that show enrichment for differential expression, we are now interested whether these findings can be supported by known regulatory interactions.</p>
<p>For example, we want to know whether transcription factors and their target genes are expressed in accordance to the connecting regulations (activation/inhibition). Such information is usually given in a gene regulatory network derived from specific experiments or compiled from the literature (<a href="https://doi.org/10.1093/nar/gkt631">Geistlinger et al., 2013</a> for an example).</p>
<p>There are well-studied processes and organisms for which comprehensive and well-annotated regulatory networks are available, e.g. the <a href="http://regulondb.ccg.unam.mx">RegulonDB</a> for <em>E. coli</em> and <a href="http://www.yeastract.com">Yeastract</a> for <em>S. cerevisiae</em>.</p>
<p>However, there are also cases where such a network is missing or at least incomplete. A basic workaround is to compile a network from regulations in pathway databases such as KEGG.</p>
<div class="sourceCode" id="cb84"><pre class="downlit">
<span class="va">hsa.grn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/compileGRN.html">compileGRN</a></span><span class="op">(</span>org <span class="op">=</span> <span class="st">"hsa"</span>, db <span class="op">=</span> <span class="st">"kegg"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">hsa.grn</span><span class="op">)</span></pre></div>
<pre><code>##      FROM    TO          TYPE
## [1,] "10000" "100132074" "-" 
## [2,] "10000" "1026"      "-" 
## [3,] "10000" "1027"      "-" 
## [4,] "10000" "10488"     "+" 
## [5,] "10000" "107"       "+" 
## [6,] "10000" "107080638" "-"</code></pre>
<p>Signaling pathway impact analysis (SPIA) is a network-based enrichment analysis method, which is explicitly designed for KEGG signaling pathways <a href="https://doi.org/%2010.1093/bioinformatics/btn577">Tarca et al., 2009</a>. The method evaluates whether expression changes are propagated across the pathway topology in combination with ORA.</p>
<div class="sourceCode" id="cb86"><pre class="downlit">
<span class="va">spia.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/nbea.html">nbea</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"spia"</span>, se <span class="op">=</span> <span class="va">allSE</span>, gs <span class="op">=</span> <span class="va">hsa.gs</span>, grn <span class="op">=</span> <span class="va">hsa.grn</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></pre></div>
<pre><code>## 
## Done pathway 1 : RNA transport..
## Done pathway 2 : RNA degradation..
## Done pathway 3 : PPAR signaling pathway..
## Done pathway 4 : Fanconi anemia pathway..
## Done pathway 5 : MAPK signaling pathway..
## Done pathway 6 : ErbB signaling pathway..
## Done pathway 7 : Calcium signaling pathway..
## Done pathway 8 : Cytokine-cytokine receptor int..
## Done pathway 9 : Chemokine signaling pathway..
## Done pathway 10 : NF-kappa B signaling pathway..
## Done pathway 11 : Phosphatidylinositol signaling..
## Done pathway 12 : Neuroactive ligand-receptor in..
## Done pathway 13 : Cell cycle..
## Done pathway 14 : Oocyte meiosis..
## Done pathway 15 : p53 signaling pathway..
## Done pathway 16 : Sulfur relay system..
## Done pathway 17 : SNARE interactions in vesicula..
## Done pathway 18 : Regulation of autophagy..
## Done pathway 19 : Protein processing in endoplas..
## Done pathway 20 : Lysosome..
## Done pathway 21 : mTOR signaling pathway..
## Done pathway 22 : Apoptosis..
## Done pathway 23 : Vascular smooth muscle contrac..
## Done pathway 24 : Wnt signaling pathway..
## Done pathway 25 : Dorso-ventral axis formation..
## Done pathway 26 : Notch signaling pathway..
## Done pathway 27 : Hedgehog signaling pathway..
## Done pathway 28 : TGF-beta signaling pathway..
## Done pathway 29 : Axon guidance..
## Done pathway 30 : VEGF signaling pathway..
## Done pathway 31 : Osteoclast differentiation..
## Done pathway 32 : Focal adhesion..
## Done pathway 33 : ECM-receptor interaction..
## Done pathway 34 : Cell adhesion molecules (CAMs)..
## Done pathway 35 : Adherens junction..
## Done pathway 36 : Tight junction..
## Done pathway 37 : Gap junction..
## Done pathway 38 : Complement and coagulation cas..
## Done pathway 39 : Antigen processing and present..
## Done pathway 40 : Toll-like receptor signaling p..
## Done pathway 41 : NOD-like receptor signaling pa..
## Done pathway 42 : RIG-I-like receptor signaling ..
## Done pathway 43 : Cytosolic DNA-sensing pathway..
## Done pathway 44 : Jak-STAT signaling pathway..
## Done pathway 45 : Natural killer cell mediated c..
## Done pathway 46 : T cell receptor signaling path..
## Done pathway 47 : B cell receptor signaling path..
## Done pathway 48 : Fc epsilon RI signaling pathwa..
## Done pathway 49 : Fc gamma R-mediated phagocytos..
## Done pathway 50 : Leukocyte transendothelial mig..
## Done pathway 51 : Intestinal immune network for ..
## Done pathway 52 : Circadian rhythm - mammal..
## Done pathway 53 : Long-term potentiation..
## Done pathway 54 : Neurotrophin signaling pathway..
## Done pathway 55 : Retrograde endocannabinoid sig..
## Done pathway 56 : Glutamatergic synapse..
## Done pathway 57 : Cholinergic synapse..
## Done pathway 58 : Serotonergic synapse..
## Done pathway 59 : GABAergic synapse..
## Done pathway 60 : Dopaminergic synapse..
## Done pathway 61 : Long-term depression..
## Done pathway 62 : Olfactory transduction..
## Done pathway 63 : Taste transduction..
## Done pathway 64 : Phototransduction..
## Done pathway 65 : Regulation of actin cytoskelet..
## Done pathway 66 : Insulin signaling pathway..
## Done pathway 67 : GnRH signaling pathway..
## Done pathway 68 : Progesterone-mediated oocyte m..
## Done pathway 69 : Melanogenesis..
## Done pathway 70 : Adipocytokine signaling pathwa..
## Done pathway 71 : Type II diabetes mellitus..
## Done pathway 72 : Type I diabetes mellitus..
## Done pathway 73 : Maturity onset diabetes of the..
## Done pathway 74 : Aldosterone-regulated sodium r..
## Done pathway 75 : Endocrine and other factor-reg..
## Done pathway 76 : Vasopressin-regulated water re..
## Done pathway 77 : Salivary secretion..
## Done pathway 78 : Gastric acid secretion..
## Done pathway 79 : Pancreatic secretion..
## Done pathway 80 : Carbohydrate digestion and abs..
## Done pathway 81 : Bile secretion..
## Done pathway 82 : Mineral absorption..
## Done pathway 83 : Alzheimer's disease..
## Done pathway 84 : Parkinson's disease..
## Done pathway 85 : Amyotrophic lateral sclerosis ..
## Done pathway 86 : Huntington's disease..
## Done pathway 87 : Prion diseases..
## Done pathway 88 : Cocaine addiction..
## Done pathway 89 : Amphetamine addiction..
## Done pathway 90 : Morphine addiction..
## Done pathway 91 : Alcoholism..
## Done pathway 92 : Bacterial invasion of epitheli..
## Done pathway 93 : Vibrio cholerae infection..
## Done pathway 94 : Epithelial cell signaling in H..
## Done pathway 95 : Pathogenic Escherichia coli in..
## Done pathway 96 : Shigellosis..
## Done pathway 97 : Salmonella infection..
## Done pathway 98 : Pertussis..
## Done pathway 99 : Legionellosis..
## Done pathway 100 : Leishmaniasis..
## Done pathway 101 : Chagas disease (American trypa..
## Done pathway 102 : African trypanosomiasis..
## Done pathway 103 : Malaria..
## Done pathway 104 : Toxoplasmosis..
## Done pathway 105 : Amoebiasis..
## Done pathway 106 : Staphylococcus aureus infectio..
## Done pathway 107 : Tuberculosis..
## Done pathway 108 : Hepatitis C..
## Done pathway 109 : Measles..
## Done pathway 110 : Influenza A..
## Done pathway 111 : HTLV-I infection..
## Done pathway 112 : Herpes simplex infection..
## Done pathway 113 : Epstein-Barr virus infection..
## Done pathway 114 : Pathways in cancer..
## Done pathway 115 : Transcriptional misregulation ..
## Done pathway 116 : Viral carcinogenesis..
## Done pathway 117 : Colorectal cancer..
## Done pathway 118 : Renal cell carcinoma..
## Done pathway 119 : Pancreatic cancer..
## Done pathway 120 : Endometrial cancer..
## Done pathway 121 : Glioma..
## Done pathway 122 : Prostate cancer..
## Done pathway 123 : Thyroid cancer..
## Done pathway 124 : Basal cell carcinoma..
## Done pathway 125 : Melanoma..
## Done pathway 126 : Bladder cancer..
## Done pathway 127 : Chronic myeloid leukemia..
## Done pathway 128 : Acute myeloid leukemia..
## Done pathway 129 : Small cell lung cancer..
## Done pathway 130 : Non-small cell lung cancer..
## Done pathway 131 : Asthma..
## Done pathway 132 : Autoimmune thyroid disease..
## Done pathway 133 : Systemic lupus erythematosus..
## Done pathway 134 : Rheumatoid arthritis..
## Done pathway 135 : Allograft rejection..
## Done pathway 136 : Graft-versus-host disease..
## Done pathway 137 : Arrhythmogenic right ventricul..
## Done pathway 138 : Dilated cardiomyopathy..
## Done pathway 139 : Viral myocarditis..</code></pre>
<pre><code>## Finished SPIA analysis</code></pre>
<div class="sourceCode" id="cb89"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/eaBrowse.html">gsRanking</a></span><span class="op">(</span><span class="va">spia.all</span><span class="op">)</span></pre></div>
<pre><code>## DataFrame with 5 rows and 6 columns
##                 GENE.SET      SIZE       NDE     T.ACT    STATUS      PVAL
##              &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## 1 hsa04620_Toll-like_r..        75         6    -3.610        -1    0.0477
## 2 hsa05416_Viral_myoca..        45         5     2.970         1    0.0733
## 3 hsa05202_Transcripti..        94         9    -0.209        -1    0.0796
## 4 hsa04630_Jak-STAT_si..        72         8    -1.060        -1    0.0908
## 5 hsa05143_African_try..        22         4     0.263         1    0.1790</code></pre>
<p>More generally applicable is gene graph enrichment analysis (GGEA), which evaluates consistency of interactions in a given gene regulatory network with the observed expression data <a href="https://doi.org/10.1093/bioinformatics/btr228">Geistlinger et al., 2011</a>.</p>
<div class="sourceCode" id="cb91"><pre class="downlit">
<span class="va">ggea.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/nbea.html">nbea</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"ggea"</span>, se <span class="op">=</span> <span class="va">allSE</span>, gs <span class="op">=</span> <span class="va">hsa.gs</span>, grn <span class="op">=</span> <span class="va">hsa.grn</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/eaBrowse.html">gsRanking</a></span><span class="op">(</span><span class="va">ggea.all</span><span class="op">)</span></pre></div>
<pre><code>## DataFrame with 8 rows and 5 columns
##                 GENE.SET   NR.RELS RAW.SCORE NORM.SCORE      PVAL
##              &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;
## 1 hsa04390_Hippo_signa..        64     23.20      0.363     0.002
## 2 hsa05416_Viral_myoca..         7      3.30      0.471     0.005
## 3 hsa05217_Basal_cell_..        18      6.92      0.385     0.013
## 4 hsa04350_TGF-beta_si..        19      7.20      0.379     0.019
## 5 hsa04910_Insulin_sig..        32     11.60      0.364     0.022
## 6 hsa04520_Adherens_ju..        13      5.02      0.386     0.025
## 7 hsa04550_Signaling_p..        36     13.10      0.363     0.025
## 8 hsa05412_Arrhythmoge..         4      1.75      0.437     0.045</code></pre>
<div class="sourceCode" id="cb93"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/nbea.html">nbeaMethods</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<pre><code>## [1] "ggea"        "spia"        "pathnet"     "degraph"     "ganpa"      
## [6] "cepa"        "topologygsa" "netgsa"      "neat"</code></pre>
<p>Note #1: As network-based enrichment methods typically do not involve sample permutation but rather network permutation, thus avoiding DE re-computation, they can likewise be applied to RNA-seq data.</p>
<p>Note #2: Given the various enrichment methods with individual benefits and limitations, combining multiple methods can be beneficial, e.g. combined application of a set-based and a network-based method. This has been shown to filter out spurious hits of individual methods and to reduce the outcome to gene sets accumulating evidence from different methods <a href="https://doi.org/10.1186/s12859-016-0884-1">Geistlinger et al., 2016</a>, <a href="https://doi.org/10.1093/bioinformatics/btw623">Alhamdoosh et al., 2017</a>.</p>
<p>The function <code>combResults</code> implements the straightforward combination of results, thereby facilitating seamless comparison of results across methods. For demonstration, we use the ORA and GSEA results for the ALL dataset from the previous section:</p>
<div class="sourceCode" id="cb95"><pre class="downlit">
<span class="va">res.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">ora.all</span>, <span class="va">gsea.all</span><span class="op">)</span>
<span class="va">comb.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/combResults.html">combResults</a></span><span class="op">(</span><span class="va">res.list</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/EnrichmentBrowser/man/eaBrowse.html">gsRanking</a></span><span class="op">(</span><span class="va">comb.res</span><span class="op">)</span></pre></div>
<pre><code>## DataFrame with 20 rows and 6 columns
##                   GENE.SET  ORA.RANK GSEA.RANK MEAN.RANK  ORA.PVAL GSEA.PVAL
##                &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## 1   hsa05202_Transcripti..       2.6       7.7       5.1    0.0351   0.00000
## 2   hsa05412_Arrhythmoge..       5.1       7.7       6.4    0.0717   0.00000
## 3   hsa04670_Leukocyte_t..      10.3      12.8      11.5    0.1220   0.00194
## 4   hsa04520_Adherens_ju..      20.5       7.7      14.1    0.2010   0.00000
## 5   hsa05100_Bacterial_i..      12.8      28.2      20.5    0.1620   0.01190
## ...                    ...       ...       ...       ...       ...       ...
## 16  hsa05410_Hypertrophi..      41.0      51.3      46.2     0.403    0.0537
## 17  hsa04350_TGF-beta_si..      38.5      59.0      48.7     0.390    0.0958
## 18  hsa04621_NOD-like_re..      61.5      38.5      50.0     0.631    0.0216
## 19  hsa04514_Cell_adhesi..      69.2      33.3      51.3     0.649    0.0158
## 20  hsa04622_RIG-I-like_..      15.4      87.2      51.3     0.178    0.3550</code></pre>
<p><em>Exercise:</em> Carry out <code>SPIA</code> and <code>GGEA</code> for the <code>airSE</code> and combine the results. How many gene sets are rendered significant by both methods?</p>
</div>
<div id="genomic-region-enrichment-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#genomic-region-enrichment-analysis" class="anchor"></a>Genomic region enrichment analysis</h2>
<p>Microarrays and next-generation sequencing are also widely applied for large-scale detection of variable and regulatory genomic regions, e.g. single nucleotide polymorphisms, copy number variations, and transcription factor binding sites.</p>
<p><img src="ENCODE.png" width="100%"></p>
<p>Such experimentally-derived genomic region sets are raising similar questions regarding functional enrichment as in gene expression data analysis.</p>
<p>Of particular interest is thereby whether experimentally-derived regions overlap more (enrichment) or less (depletion) than expected by chance with regions representing known functional features such as genes or promoters.</p>
<p>The <a href="http://bioconductor.org/packages/regioneR">regioneR</a> package implements a general framework for testing overlaps of genomic regions based on permutation sampling. This allows to repeatedly sample random regions from the genome, matching size and chromosomal distribution of the region set under study. By recomputing the overlap with the functional features in each permutation, statistical significance of the observed overlap can be assessed.</p>
<div class="sourceCode" id="cb97"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">regioneR</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19.masked</span><span class="op">)</span></pre></div>
<p>To demonstrate the basic functionality of the package, we consider the overlap of gene promoter regions and CpG islands in the human genome. We expect to find an enrichment as promoter regions are known to be GC-rich. Hence, is the overlap between CpG islands and promoters greater than expected by chance?</p>
<p>We use the collection of CpG islands described in <a href="https://doi.org/10.1093/biostatistics/kxq005">Wu et al., 2010</a> and restrict them to the set of canonical chromosomes 1-23, <em>X</em>, and <em>Y</em>.</p>
<div class="sourceCode" id="cb98"><pre class="downlit">
<span class="va">cpgHMM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"http://www.haowulab.org/software/makeCGI/model-based-cpg-islands-hg19.txt"</span><span class="op">)</span>
<span class="va">cpgHMM</span> <span class="op">&lt;-</span> <span class="fu">makeGRangesFromDataFrame</span><span class="op">(</span><span class="va">cpgHMM</span>, keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">genome</span><span class="op">(</span><span class="va">cpgHMM</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"hg19"</span>
<span class="va">cpgHMM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/regioneR/man/filterChromosomes.html">filterChromosomes</a></span><span class="op">(</span><span class="va">cpgHMM</span>, chr.type <span class="op">=</span> <span class="st">"canonical"</span><span class="op">)</span>
<span class="va">cpgHMM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">cpgHMM</span><span class="op">)</span>
<span class="va">cpgHMM</span></pre></div>
<pre><code>## GRanges object with 63705 ranges and 5 metadata columns:
##           seqnames            ranges strand |    length  CpGcount GCcontent
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##       [1]     chr1       10497-11241      * |       745       110       549
##       [2]     chr1       28705-29791      * |      1087       115       792
##       [3]     chr1     135086-135805      * |       720        42       484
##       [4]     chr1     136164-137362      * |      1199        71       832
##       [5]     chr1     137665-138121      * |       457        22       301
##       ...      ...               ...    ... .       ...       ...       ...
##   [63701]     chrY 59213702-59214290      * |       589        43       366
##   [63702]     chrY 59240512-59241057      * |       546        40       369
##   [63703]     chrY 59348047-59348370      * |       324        17       193
##   [63704]     chrY 59349137-59349565      * |       429        31       276
##   [63705]     chrY 59361489-59362401      * |       913       128       650
##               pctGC    obsExp
##           &lt;numeric&gt; &lt;numeric&gt;
##       [1]     0.737     1.106
##       [2]     0.729     0.818
##       [3]     0.672     0.548
##       [4]     0.694     0.524
##       [5]     0.659     0.475
##       ...       ...       ...
##   [63701]     0.621     0.765
##   [63702]     0.676     0.643
##   [63703]     0.596     0.593
##   [63704]     0.643     0.700
##   [63705]     0.712     1.108
##   -------
##   seqinfo: 24 sequences from hg19 genome; no seqlengths</code></pre>
<p>Analogously, we load promoter regions in the <em>hg19</em> human genome assembly as available from <a href="https://genome.ucsc.edu/">UCSC</a>:</p>
<div class="sourceCode" id="cb100"><pre class="downlit">
<span class="va">promoters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"http://gattaca.imppc.org/regioner/data/UCSC.promoters.hg19.bed"</span><span class="op">)</span>
<span class="va">promoters</span> <span class="op">&lt;-</span> <span class="fu">makeGRangesFromDataFrame</span><span class="op">(</span><span class="va">promoters</span>, keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">genome</span><span class="op">(</span><span class="va">promoters</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"hg19"</span>
<span class="va">promoters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/regioneR/man/filterChromosomes.html">filterChromosomes</a></span><span class="op">(</span><span class="va">promoters</span>, chr.type <span class="op">=</span> <span class="st">"canonical"</span><span class="op">)</span>
<span class="va">promoters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">promoters</span><span class="op">)</span>
<span class="va">promoters</span></pre></div>
<pre><code>## GRanges object with 49049 ranges and 3 metadata columns:
##           seqnames            ranges strand |          V4          V5
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
##       [1]     chr1        9873-12073      * |         TSS           .
##       [2]     chr1       16565-18765      * |         TSS           .
##       [3]     chr1       17551-19751      * |         TSS           .
##       [4]     chr1       17861-20061      * |         TSS           .
##       [5]     chr1       19559-21759      * |         TSS           .
##       ...      ...               ...    ... .         ...         ...
##   [49045]     chrY 59211948-59214148      * |         TSS           .
##   [49046]     chrY 59328251-59330451      * |         TSS           .
##   [49047]     chrY 59350972-59353172      * |         TSS           .
##   [49048]     chrY 59352984-59355184      * |         TSS           .
##   [49049]     chrY 59360654-59362854      * |         TSS           .
##                    V6
##           &lt;character&gt;
##       [1]           +
##       [2]           -
##       [3]           -
##       [4]           -
##       [5]           -
##       ...         ...
##   [49045]           +
##   [49046]           +
##   [49047]           +
##   [49048]           +
##   [49049]           -
##   -------
##   seqinfo: 24 sequences from hg19 genome; no seqlengths</code></pre>
<p>To speed up the example, we restrict analysis to chromosomes 21 and 22. Note that this is done for demonstration only. To make an accurate claim, the complete region set should be used (which, however, runs considerably longer).</p>
<div class="sourceCode" id="cb102"><pre class="downlit">
<span class="va">cpg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">cpgHMM</span>, <span class="va">seqnames</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr21"</span>, <span class="st">"chr22"</span><span class="op">)</span><span class="op">)</span>
<span class="va">prom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">promoters</span>, <span class="va">seqnames</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr21"</span>, <span class="st">"chr22"</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>Now, we are applying an overlap permutation test with 100 permutations (<code>ntimes = 100</code>), while maintaining chromosomal distribution of the CpG island region set (<code>per.chromosome = TRUE</code>). Furthermore, we use the option <code>count.once = TRUE</code> to count an overlapping CpG island only once, even if it overlaps with 2 or more promoters.</p>
<p>Note that we use 100 permutations for demonstration only. To draw robust conclusions a minimum of 1000 permutations should be carried out.</p>
<div class="sourceCode" id="cb103"><pre class="downlit">
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/regioneR/man/overlapPermTest.html">overlapPermTest</a></span><span class="op">(</span><span class="va">cpg</span>, <span class="va">prom</span>, mask <span class="op">=</span> <span class="cn">NA</span>,
                       genome <span class="op">=</span> <span class="st">"hg19"</span>, ntimes <span class="op">=</span> <span class="fl">100</span>,
                       per.chromosome <span class="op">=</span> <span class="cn">TRUE</span>, count.once <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">res</span></pre></div>
<pre><code>## $numOverlaps
## P-value: 0.0099009900990099
## Z-score: 56.9304
## Number of iterations: 100
## Alternative: greater
## Evaluation of the original region set: 719
## Evaluation function: numOverlaps
## Randomization function: randomizeRegions
## 
## attr(,"class")
## [1] "permTestResultsList"</code></pre>
<div class="sourceCode" id="cb105"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">permuted</span><span class="op">)</span></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##    86.0   112.0   117.0   117.4   123.0   139.0</code></pre>
<p>The resulting permutation <em>p</em>-value indicates a significant enrichment. Out of the 2859 CpG islands, 719 overlap with at least one promoter. In contrast, when repeatedly drawing random regions matching the CpG islands in size and chromosomal distribution, the mean number of overlapping regions across permutations was 117.4 <span class="math inline">\(\pm\)</span> 10.6.</p>
<p>Note #1: The function <code><a href="https://rdrr.io/pkg/regioneR/man/permTest.html">regioneR::permTest</a></code> allows to incorporate user-defined functions for randomizing regions and evaluating additional measures of overlap such as total genomic size in bp.</p>
<p>Note #2: The <a href="http://bioconductor.org/packages/LOLA">LOLA</a> package implements a genomic region ORA, which assesses genomic region overlap based on the hypergeometric distribution using a library of pre-defined functional region sets.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
